### PR TITLE
Fix another set of minor issues

### PR DESCRIPTION
**Pull request by desilinguist**
- Ensure that no experiment ID can contain spaces (along with tests)
- Check if the optional columns for `rsmeval` are missing in the predictions file (along with tests)
- Add docstrings to all functions in all files.

### CODE REVIEW

File: rsmtool/analysis.py [hunk 1]
```diff
@@ -95,7 +123,21 @@ def compute_outliers(df, selected_features):
     """
     Compute the number and percentage of outliers
     outside mean +/- 4 SD for the given columns with
```
Comment (aloukina) [inline rsmtool/analysis.py]: Delete 'with'?

File: rsmtool/analysis.py [hunk 2]
```diff
@@ -368,17 +449,43 @@ def compute_metrics(df,
...
+
+    Parameters
+    ----------
+    df : pandas DataFrame
+        Input data frame
+    compute_shortened : bool, optional
+        Also compute a shortened version of the full
+        metrics data frame, defaults to False.
+    use_scaled_predictions : bool, optional
+        Include scaled predictions for evaluation
```
Comment (aloukina) [inline rsmtool/analysis.py]: Change to "Use evaluations based on scaled predictions in the shortened version of the metrics data frame"

We always compute full evaluations for both raw and scale  as long as they are supplied in the data.

File: rsmtool/analysis.py [hunk 3]
```diff
@@ -448,9 +555,34 @@ def compute_metrics_by_group(df_test,
...
+
+    Parameters
+    ----------
+    df_test : pandas DataFrame
+        Input data frame.
+    grouping_variable : str
+        Feature name indicating the column that
+        contains grouping information.
+    use_scaled_predictions : bool, optional
+        Include scaled predictions when computing
```
Comment (aloukina) [inline rsmtool/analysis.py]: "correlations statistics" -> "evaluation metrics" (not just correlation. Scaling has no effect on correlations anyway)

File: rsmtool/analysis.py [hunk 4]
```diff
@@ -448,9 +555,34 @@ def compute_metrics_by_group(df_test,
...
+        the correlation statistics, defaults to False.
+    include_second_score : bool, optional
```
Comment (aloukina) [inline rsmtool/analysis.py]: Correlation -> association

File: rsmtool/analysis.py [hunk 5]
```diff
@@ -448,9 +555,34 @@ def compute_metrics_by_group(df_test,
...
+        Include human-human correlation statistics,
+        defaults to False.
+
+    Returns
+    -------
+    df_human_machine_eval_by_group : pandas DataFrame
+        Data frame containing the correlation
+        human-machine correlation statistics.
```
Comment (aloukina) [inline rsmtool/analysis.py]: correlation -> association

File: rsmtool/analysis.py [hunk 6]
```diff
@@ -448,9 +555,34 @@ def compute_metrics_by_group(df_test,
...
+    df_human_human_eval_by_group : pandas DataFrame
+        Dat frame that either contains the human-human
```
Comment (aloukina) [inline rsmtool/analysis.py]: Dat -> data

File: rsmtool/analysis.py [hunk 7]
```diff
@@ -495,12 +627,27 @@ def compute_metrics_by_group(df_test,
...
+    use_all_responses : bool, optional
+        Use the full data set instead of only using
+        the double-scored subset, defaults to True.
+
+    Returns
+    -------
+    df_degradation : pandas DataFrame
+        Data frame containing the degradation
+        statistics pandas DataFrame,
     """
```
Comment (aloukina) [inline rsmtool/analysis.py]: Remove second 'pandas DataFrame'?

File: rsmtool/analysis.py [hunk 8]
```diff
@@ -630,10 +804,33 @@ def run_prediction_analyses(df_test,
...
+        Data frame containing the human scores
+        for the test set.
+    subgroups : list of str
+        List of columns containing grouping information.
+    second_human_score_column : str
+        Column name that contains the second human scores.
+    use_scaled_predictions : bool, optional
+        Include scaled predictions when computing
+        computing the evaluation statistics, defaults
+        to False.
```
Comment (aloukina) [inline rsmtool/analysis.py]: "Use scaled predictions for computing in-depth analyses"

File: rsmtool/analysis.py [hunk 9]
```diff
@@ -816,8 +1084,43 @@ def run_data_composition_analyses_for_rsmtool(df_train_metadata,
...
+
+    Parameters
+    ----------
+    df_train_metadata : pandas DataFrame
+        Data frame containing the metadata information
+        about the responses in the training set.
+    df_test_metadata : pandas DataFrame
+        Data frame containing the metadata information
+        about the responses in the test set.
+    df_train_excluded : data frame containing the responses
```
Comment (aloukina) [inline rsmtool/analysis.py]: pandas DataFrame (for consistency)

File: rsmtool/input.py [hunk 10]
```diff
@@ -51,7 +71,21 @@ def locate_custom_sections(custom_report_section_paths, configpath):
 def normalize_json_fields(json_obj):
     """
     Normalize the the field names in `json_obj` in order to
```
Comment (aloukina) [inline rsmtool/input.py]: Delete one 'the'

File: rsmtool/input.py [hunk 11]
```diff
@@ -249,11 +314,25 @@ def validate_and_populate_json_fields(json_obj, context='rsmtool'):
 
 def process_json_fields(json_obj):
     """
-    Converts json fields which are read in
-    string to an appropriate format. Fields which
-    can take multiple values lists are converted to lists
-    if they have not been already formatted as such.
+    Converts fields which are read in as string to the
+    appropriate format. Fields which can take multiple values
+    lists are converted to lists if they have not been already
```
Comment (aloukina) [inline rsmtool/input.py]: I feel like a prepositions is missing between values and lists

File: rsmtool/input.py [hunk 12]
```diff
@@ -292,24 +384,44 @@ def parse_json_with_comments(filename):
...
+        return ans
 
 
 def normalize_and_validate_feature_file(feature_json):
-
     """
-    Normalize the the field names in `feature_json`
-    in order to maintain backwards compatibility
-    with old config files.
+    Normalize the the field names in `feature_json` in order to maintain
```
Comment (aloukina) [inline rsmtool/input.py]: the the

File: rsmtool/input.py [hunk 13]
```diff
@@ -360,10 +490,28 @@ def read_json_file(json_file):
 def check_main_config(obj, context='rsmtool'):
 
     """
-    Normalize all fields in main config file, check for
-    all required fields, add the default values, and convert
-    all strings to appropriate objects.
+    The driver function that creates the final JSON object from the
+    configuration file. Normalizes all fields in main config file, checks
+    for all required fields, adds the default values, and convert all
```
Comment (aloukina) [inline rsmtool/input.py]: convertS

File: rsmtool/input.py [hunk 14]
```diff
@@ -401,9 +561,27 @@ def locate_file(filepath, configpath):
 def check_subgroups(df, subgroups):
 
     """
-    Check that all subgroups, if specified, correspond to
-    columns in the provided data frame, and replace all NaNs
-    in subgroups values with 'No info' for later convenience.
+    Check that all subgroups, if specified, correspond to columns in the
+    provided data frame, and replace all NaNs in subgroups values with
+    'No info' for later convenience. Raises an excetion if any specified
```
Comment (aloukina) [inline rsmtool/input.py]: excePtion

File: rsmtool/model.py [hunk 15]
```diff
@@ -42,14 +43,27 @@
...
+
 def ols_coefficients_to_dataframe(coefs):
     """
-    Take the given series with feature names in the index and
-    the coefficient values as the data, obtained from a linear model
-    trained using `pandas.ols` and convert it to a data frame with
-    two columns, the first being the feature name and the second being
-    the coefficient value. The first row in the output data frame is
-    always for the intercept and the rest are sorted by feature name.
+    Convert a series containing OLS coefficients and convert it
```
Comment (aloukina) [inline rsmtool/model.py]: Convert a series .. coefficients to a data frame
or Take a series...

File: rsmtool/model.py [hunk 16]
```diff
@@ -42,14 +43,27 @@
...
+    to a data frame.
+
+    Parameters
+    ----------
+    coefs : `pandas` Series
```
Comment (aloukina) [inline rsmtool/model.py]: We don't use back quotes around `pandas` in other doc strings

File: rsmtool/model.py [hunk 17]
```diff
@@ -72,14 +86,26 @@ def ols_coefficients_to_dataframe(coefs):
...
+    at the coefficients and the intercept because the
+    `model_params` attribute of the SKLL model ignores
+    zero coefficients, which we do not want. The first
+    row in the output data frame is always for the
     intercept and the rest are sorted by feature name.
-    """
 
+    Parameters
+    ----------
+    learner : `skll` Learner object
```
Comment (aloukina) [inline rsmtool/model.py]: As above (and also line 105)

File: rsmtool/preprocess.py [hunk 18]
```diff
@@ -15,6 +15,22 @@ def trim(values, trim_min, trim_max):
     """
     Trim the values contained in the given vector
     `values` to `trim_min` and `trim_max`
```
Comment (aloukina) [inline rsmtool/preprocess.py]: We should probably edit this for accuracy to refer to trim_max+0.49998 and trim_min-0.49998

File: rsmtool/preprocess.py [hunk 19]
```diff
@@ -177,99 +263,230 @@ def remove_outliers(data, mean=None, sd=None, sd_multiplier=4):
...
     """
     Apply the inverse transform to `data`.
+
+    Parameters
+    ----------
+    name : str
+        Name of the feature to transform.
+    data : numpy array
+        Numpy array containing the feature values.
+    sd_multiplier : int, optional
```
Comment (aloukina) [inline rsmtool/preprocess.py]: "Use this std. dev. multiplier to compute the ceiling and floor for outlier removal and check that these are not equal to zero"

File: rsmtool/preprocess.py [hunk 20]
```diff
@@ -177,99 +263,230 @@ def remove_outliers(data, mean=None, sd=None, sd_multiplier=4):
...
+    new_data: numpy array
+        Numpy array containing the transformed feature
+        values.
+
+    Raises
+    ------
+    ValueError
+        If the transform is applied to a feature that can
+        be zero or to a feature that can be either positive
+        or negative.
```
Comment (aloukina) [inline rsmtool/preprocess.py]: either -> both ?

File: rsmtool/preprocess.py [hunk 21]
```diff
@@ -177,99 +263,230 @@ def remove_outliers(data, mean=None, sd=None, sd_multiplier=4):
...
+    -------
+    new_data : numpy array
+        Numpy array containing the transformed feature
+        values.
+
+    Raises
+    ------
+    ValueError
+        If the transform is applied to a feature
+        that can be zero.
```
Comment (aloukina) [inline rsmtool/preprocess.py]: can be zero -> has negative values

File: rsmtool/preprocess.py [hunk 22]
```diff
@@ -282,16 +499,51 @@ def transform_feature(name, data, transform):
...
+    data : numpy array
+        Numpy array containing the values to pre-process.
+    feature_name : str
+        Name of the feature being pre-processed.
+    feature_transform : str
+        Name of the transformation function to apply.
+    feature_mean : float
+        Mean value to use for outlier detection instead
+        of the mean of the given feature values.
+    feature_sd : TYPE
```
Comment (aloukina) [inline rsmtool/preprocess.py]: ?

File: rsmtool/preprocess.py [hunk 23]
```diff
@@ -282,16 +499,51 @@ def transform_feature(name, data, transform):
...
+        Std. dev. value to use for outlier detection instead
+        of the std. dev. of the given feature values.
+    exclude_zero_sd : bool, optional
+        Check whether there are features that have a zero
```
Comment (aloukina) [inline rsmtool/preprocess.py]: "Exclude features which have a zero standard deviation"

File: rsmtool/rsmeval.py [hunk 24]
```diff
@@ -382,7 +390,6 @@ def run_evaluation(config_file, output_dir):
                                               df_test_human_scores,
                                               subgroups,
                                               second_human_score_column,
-                                              exclude_zero_scores=exclude_zero_scores,
```
Comment (aloukina) [inline rsmtool/rsmeval.py]: Why?

File: rsmtool/utils.py [hunk 25]
```diff
@@ -72,14 +89,22 @@ def covariance_to_correlation(m):
...
+    df : pandas DataFrame
+        Data frame containing the feature values.
+
+    Returns
+    -------
+    df_pcor : pandas DataFrame
+        Data frame containing the partial correlations.
     """
     numrows, numcols = df.shape
-    gp = numcols - 2
```
Comment (aloukina) [inline rsmtool/utils.py]: Why?

File: rsmtool/utils.py [hunk 26]
```diff
@@ -123,17 +157,32 @@ def agreement(score1, score2, tolerance=0):
     return (float(num_agreements) / len(score1)) * 100
```
Comment (aloukina) [inline rsmtool/utils.py]: Elsewhere these were changed to
agreement = float(num_agreements)....
return agreement

**Comment by aloukina** [comment]
I am done reviewing. Most of my comments are really nitpicks apart a couple of corrections to docstrings and the cases where I am not sure what the change is doing.

File: rsmtool/rsmeval.py (hunk 24 above)
Comment (desilinguist) [inline rsmtool/rsmeval.py]: It was removed because this flag is not used at all inside `run_prediction_analyses()`.

**Pull request merged**